File: IOS/ChordRecognizeGenerate/ChordRecognizeGenerate/ChordGenerator.swift
//
//  ChordGenerator.swift
//  ChordRecognizeGenerate
//
//  Created by Derrick Ward on 12/20/19.
//  Copyright © 2019 Derrick Ward. All rights reserved.
//

import Foundation

@objc public class ChordGenerator : NSObject
{
    
    public override init() {}
    
    @objc public func chordToMIDINotes(chord: Chord) -> [ChordNote]
    {
        var chordNotes = [NoteRootOffset]()
        
        switch chord.quality
        {
            case Quality.Maj:
                ChordDefs.majTriad.addNotes(chordNotes: &chordNotes,inversion: chord.inversion)
                chord.addNotesFromOffsets(offsets: chordNotes)
                chord.triad = ChordDefs.majTriad
            case Quality.min:
                ChordDefs.minTriad.addNotes(chordNotes: &chordNotes,inversion: chord.inversion)
                chord.addNotesFromOffsets(offsets: chordNotes)
                chord.triad = ChordDefs.minTriad
            case Quality.aug:
                ChordDefs.augTriad.addNotes(chordNotes: &chordNotes,inversion: chord.inversion)
                chord.addNotesFromOffsets(offsets: chordNotes)
                chord.triad = ChordDefs.augTriad
            case Quality.dim:
                 ChordDefs.dimTriad.addNotes(chordNotes: &chordNotes,inversion: chord.inversion)
                 chord.addNotesFromOffsets(offsets:chordNotes)
                 chord.triad = ChordDefs.dimTriad
            case Quality.sus2:
                ChordDefs.sus2Triad.addNotes(chordNotes: &chordNotes,inversion: chord.inversion)
                chord.addNotesFromOffsets(offsets: chordNotes)
                chord.triad = ChordDefs.sus2Triad
            case Quality.sus4:
                ChordDefs.sus4Triad.addNotes(chordNotes: &chordNotes,inversion: chord.inversion)
                chord.addNotesFromOffsets(offsets: chordNotes)
                chord.triad = ChordDefs.sus4Triad
            default:
                ChordDefs.majTriad.addNotes(chordNotes: &chordNotes,inversion: chord.inversion)
                chord.addNotesFromOffsets(offsets: chordNotes)
                chord.triad = ChordDefs.majTriad
        }
        
        parseFactors(chord: chord)
               parseAlterations(chord: chord)
        parseAdditions(chord: chord)
        
        if chord.ommission != nil
        {
            let ommissionOffset = chord.ommission! == Degree.five ? chord.triad.fifth : chord.triad.third
            let note : ChordNote = chord.rootNote! + ChordNote(ommissionOffset)
            chord.notes.removeAll(where: { $0 == note })
        }
        
        let adjustedRoot = chord.inversion != .none ? chord.rootNote! + 12 : chord.rootNote!
        
        chord.notes.forEach({
            if $0 < adjustedRoot
            {
                chord.invertedNotes.append($0)
            }
        })
        
        return chord.notes
    }
    
    private func parseAdditions(chord: Chord)
    {
        chord.additions.forEach({ chord.notes.append($0.toNote(root: chord.rootNote!)) })
    }
    
    private func parseFactors(chord : Chord)
    {
        let tones = chord.factors
        let seventhOffsetVal = seventhOffset(chordQuality: chord.quality, factorQuality: chord.factorQuality)
        
        for tone in tones
        {
            var offsets = [NoteRootOffset]()
            //special case for "sixth" chords
            if tone.degree == Degree.six || tone.degree == Degree.sixNine
            {
                let extOffsets = ChordDefs.sixthChordDesignatorToOffset(designator: tone.degree)
                offsets.append(contentsOf: extOffsets)
                chord.addNotesFromOffsets(offsets: offsets)
                continue
            }
            
            let sigOffset = ChordDefs.signatureOffset(designators: tone.signs)
            guard let extensionOffset = ChordDefs.degreeToOffset[tone.degree] else { return }
            
            let degreeInScale = ChordDefs.offsetToMajDegree[Int(extensionOffset)].degree
            let numberDegreeInScale = degreeInScale.toNumber()
            if numberDegreeInScale >= 7 && numberDegreeInScale % 2 == 1
            {
                offsets.append(ChordDefs.majScaleToOffset[7] + seventhOffsetVal)
                
                for note in stride(from: 9, through: numberDegreeInScale, by: 2)
                {
                    offsets.append(ChordDefs.majScaleToOffset[Int(note)] + (note == numberDegreeInScale ? sigOffset : 0))
                }
            }
            
            chord.addNotesFromOffsets(offsets: offsets)
        }
    }
    
    private func parseAlterations(chord : Chord)
    {
        if chord.alteredNotes.count > 0
        {
            var idx = 0
            var alteredTone = chord.alteredNotes[0]
            for j in 0..<chord.notes.count
            {
                var note = chord.notes[j]
                let offset = ChordDefs.offsetInExtendedChord(rootNote: chord.rootNote!,note: note)
                let degree = ChordDefs.offsetToMajDegree[Int(offset)].degree
                if degree == alteredTone.degree
                {
                    note = ChordNote(Int8(note) + ChordDefs.signatureOffset(designators: alteredTone.signs))
                    chord.notes[j] = note
                    idx += 1
                    if idx < chord.alteredNotes.count
                    {
                        alteredTone = chord.alteredNotes[idx]
                    }
                }
            }
        }
    }
    
    private func seventhOffset(chordQuality: Quality, factorQuality : Quality) -> Int8
    {
        var offset : Int8 = -1
        
        if chordQuality == Quality.dim
        {
            offset = -2
        }
        
        switch factorQuality {
            case Quality.Maj:
                offset = 0
            //case Designator.dom:
            case Quality.halfDim:
                offset = -1
           // case Designator.min:
            //case Designator.aug:
            default:
                print("none")
        }
        
        return offset
    }

}

File: Android/ChordRecGenLib/src/main/java/com/blindintuition/chordrecgenlib/ChordGenerator.kt
package com.blindintuition.chordrecgenlib

class ChordGenerator {

    fun chordToMIDINotes(chord: Chord) : Array<ChordNote>
    {
        val chordNotes = arrayListOf<NoteRootOffset>()

        when(chord.quality) {
            Quality.Maj -> {
                ChordDefs.majTriad.addNotes(chordNotes, chord.inversion)
                chord.addNotesFromOffsets(chordNotes.toTypedArray())
                chord.triad = ChordDefs.majTriad
            }
            Quality.min -> {
                ChordDefs.minTriad.addNotes(chordNotes, chord.inversion)
                chord.addNotesFromOffsets(chordNotes.toTypedArray())
                chord.triad = ChordDefs.minTriad
            }

            Quality.aug -> {
                ChordDefs.augTriad.addNotes(chordNotes, chord.inversion)
                chord.addNotesFromOffsets(chordNotes.toTypedArray())
                chord.triad = ChordDefs.augTriad
            }
            Quality.dim -> {
                ChordDefs.dimTriad.addNotes(chordNotes, chord.inversion)
                chord.addNotesFromOffsets(chordNotes.toTypedArray())
                chord.triad = ChordDefs.dimTriad
            }
            Quality.sus2 -> {
                ChordDefs.sus2Triad.addNotes(chordNotes,chord.inversion)
                chord.addNotesFromOffsets(chordNotes.toTypedArray())
                chord.triad = ChordDefs.sus2Triad
            }
            Quality.sus4 -> {
                ChordDefs.sus4Triad.addNotes(chordNotes, chord.inversion)
                chord.addNotesFromOffsets(chordNotes.toTypedArray())
                chord.triad = ChordDefs.sus4Triad
            }
            else -> {
                ChordDefs.majTriad.addNotes(chordNotes, chord.inversion)
                chord.addNotesFromOffsets(chordNotes.toTypedArray())
                chord.triad = ChordDefs.majTriad
            }
        }

        parseFactors(chord)
        parseAlterations(chord)
        parseAdditions(chord)

        if (chord.ommission != null)
        {
            val ommissionOffset = if(chord.ommission == Degree.five)  chord.triad.fifth else chord.triad.third
            val note : ChordNote = (chord.rootNote + ommissionOffset).toChordNote()
            chord.notes.removeAll({ it == note })
        }

        val adjustedRoot = if(chord.inversion != Inversion.none) chord.rootNote + 12 else chord.rootNote

        chord.notes.forEach {
            if(it < adjustedRoot as ChordNote) {
                chord.invertedNotes.add(it)
            }
        }

        return chord.notes.toTypedArray()
    }

    private fun parseAdditions(chord: Chord)
    {
        chord.additions.forEach({ chord.notes.add(it.toNote(chord.rootNote)) })
    }

    private fun parseFactors(chord : Chord)
    {
        val tones = chord.factors
        val seventhOffsetVal = seventhOffset(chord.quality, chord.factorQuality)

        for (tone in tones)
        {
            val offsets = arrayListOf<NoteRootOffset>()
            //special case for "sixth" chords
            if (tone.degree == Degree.six || tone.degree == Degree.sixNine)
            {
                val extOffsets = ChordDefs.sixthChordDesignatorToOffset(tone.degree)
                offsets.addAll(extOffsets)
                chord.addNotesFromOffsets(offsets.toTypedArray())
                continue
            }

            val sigOffset = ChordDefs.signatureOffset(tone.signs.toTypedArray())
            val extensionOffset = ChordDefs.degreeToOffset[tone.degree]

            if(extensionOffset == null)
            {
                return
            }

            val degreeInScale = ChordDefs.offsetToMajDegree[extensionOffset.toInt()].degree
            val numberDegreeInScale = degreeInScale.number
            if ((numberDegreeInScale >= 7.toUByte()) && (numberDegreeInScale.toInt() % 2 == 1))
            {
                offsets.add((ChordDefs.majScaleToOffset[7] + seventhOffsetVal).toNoteRootOffset())

                for (note in 9..numberDegreeInScale.toInt() step 2)
                {
                    offsets.add( (ChordDefs.majScaleToOffset[note] +
                            (if(note == numberDegreeInScale.toInt()) sigOffset else 0)).toNoteRootOffset())
                }
            }

            chord.addNotesFromOffsets(offsets.toTypedArray())
        }
    }

    private fun parseAlterations(chord : Chord)
    {
        if(chord.alteredNotes.size > 0)
        {
            var idx = 0
            var alteredTone = chord.alteredNotes[0]
            for (j in 0 until chord.notes.size)
                    {
                        var note = chord.notes[j]
                        val offset = ChordDefs.offsetInExtendedChord(chord.rootNote,note)
                        val degree = ChordDefs.offsetToMajDegree[offset.toInt()].degree
                        if (degree == alteredTone.degree)
                        {
                            note = (note + ChordDefs.signatureOffset(alteredTone.signs.toTypedArray())).toChordNote()
                            chord.notes[j] = note
                            idx += 1
                            if (idx < chord.alteredNotes.size)
                            {
                                alteredTone = chord.alteredNotes[idx]
                            }
                        }
                    }
        }
    }

    private fun seventhOffset(chordQuality: Quality, factorQuality : Quality) : Byte
    {
        var offset : Byte = -1

        if (chordQuality == Quality.dim)
        {
            offset = -2
        }

        when(factorQuality)
        {
            Quality.Maj -> { offset = 0 }

            //case Designator.dom:
            Quality.halfDim -> { offset = -1 }

            // case Designator.min:
            //case Designator.aug:
            else -> {}
        }

        return offset
    }

}